packages/rocketchat-importer-hipchat-enterprise/server/importer.js

```diff
@@ -816,64 +868,90 @@ export class HipChatEnterpriseImporter extends Base {
 			console.error(e);
 			this.addMessageError(e, msg);
 		}
-
-		super.addCountCompleted(1);
 	}
 
-	_importMessages(startedByUserId) {
-		const messageListIds = this.collection.find({
-			import: this.importRecord._id,
-			importer: this.name,
-			type: 'messages',
-		}, { _id : true }).fetch();
+	async _importMessageList(startedByUserId, messageListId) {
+		const list = this.collection.findOneById(messageListId);
+		if (!list) {
+			return;
+		}
 
-		messageListIds.forEach((item) => {
-			const list = this.collection.findOneById(item._id);
-			if (!list) {
-				return;
-			}
+		if (!list.messages) {
+			return;
+		}
 
-			if (!list.messages) {
-				return;
-			}
+		const { roomIdentifier, hipchatRoomId, name } = list;
+		const rid = await this._getRoomRocketId(hipchatRoomId);
 
-			const { roomIdentifier, hipchatRoomId, name } = list;
-			const rid = this._getRoomRocketId(hipchatRoomId);
+		// If there's no rocketId for the channel, then it wasn't imported
+		if (!rid) {
+			this.logger.debug(`Ignoring room ${ roomIdentifier } ( ${ name } ), as there's no rid to use.`);
+			return;
+		}
 
-			// If there's no rocketId for the channel, then it wasn't imported
-			if (!rid) {
-				this.logger.debug(`Ignoring room ${ roomIdentifier } ( ${ name } ), as there's no rid to use.`);
-				return;
-			}
+		const room = await RocketChat.models.Rooms.findOneById(rid, { fields: { usernames: 1, t: 1, name: 1 } });
+		await super.updateRecord({
+			messagesstatus: `${ roomIdentifier }.${ list.messages.length }`,
+			'count.completed': this.progress.count.completed,
+		});
 
-			const room = RocketChat.models.Rooms.findOneById(rid, { fields: { usernames: 1, t: 1, name: 1 } });
-			super.updateRecord({
-				messagesstatus: `${ roomIdentifier }.${ list.messages.length }`,
-				'count.completed': this.progress.count.completed,
-			});
+		await Meteor.runAsUser(startedByUserId, async() => {
+			let msgCount = 0;
+			try {
+				for (const msg of list.messages) {
+					await this._importSingleMessage(msg, roomIdentifier, room);
+					msgCount++;
+					if (msgCount >= 50) {
+						super.addCountCompleted(msgCount);
+						msgCount = 0;
+					}
+				}
+			} catch (e) {
+				this.logger.error(e);
+			}
 
-			Meteor.runAsUser(startedByUserId, () => {
-				list.messages.forEach((msg) => {
-					this._importSingleMessage(msg, roomIdentifier, room);
-				});
-			});
+			if (msgCount > 0) {
+				super.addCountCompleted(msgCount);
+			}
 		});
+
+	}
+
+	async _importMessages(startedByUserId) {
+		const messageListIds = this.collection.find({
+			import: this.importRecord._id,
+			importer: this.name,
+			type: 'messages',
+		}, { fields: { _id: true } }).fetch();
+
+		for (const item of messageListIds) {
+			await this._importMessageList(startedByUserId, item._id);
+		}
 	}
 
 	_importDirectMessages() {
```

Comment:
Shouldn't this be async?

Reply:
It's sync to avoid working on too many lists at the same time, as that caused the process to use too much RAM.